### ISSUE TITLE
Update Quick Info and Method Info to show exception reasons

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#1515]
Currently the Quick Info and Method Info features for the C# editor show the types of exceptions potentially thrown by the method (according to the documentation), but it does not include the descriptions of the reasons why these exceptions could be thrown.

It would be very helpful if the documentation presentation were updated to include this information, so method callers are more likely to correctly consider preconditions.

**Comment by Pilchie** [comment]
We should take some care that Quick Info doesn't become so big that it is intrusive by adding a ton of info there.

**Comment by AnthonyDGreen** [comment]
Yeah, just ran into this myself. 
![image](https://cloud.githubusercontent.com/assets/10539978/24775197/a52457a6-1ad0-11e7-91ed-57c50f6c13d5.png)

All that empty space next to the exceptions seems kind of a waste and without the reasons the list of exceptions is less useful. But that's just my first impression. I'll keep using it and see if I feel that way.

**Comment by CyrusNajmabadi** [comment]
I agree with kevin.  I'd be very worried by how cluttered this could become.  It's interesting to me that we haven't gotten any feedback over this over the years.  If we were to do something like this, I would likely want a way to collapse this so that the main meat of the quickinfo weren't drowned out by all the information around exceptions.

**Comment by AnthonyDGreen** [comment]
@CyrusNajmabadi 

I can't speak for Craig but prior to VS2015 VB Quick Info didn't display this information at all. And I don't usually use Quick Info in this way. And to my knowledge Roslyn much of the Roslyn APIs (and certainly the internals) don't provide any exception info or if they do they provide types but not reasons. This particular method both has exception type info and reason info _and_ it throws 9 exceptions making the list a _lot_ more noticeable than it normally is.

In other words, this is the first time I noticed the feature at all. And once I noticed the feature I noticed how it fell just short of providing the value I expected as I still had to hit F1 to figure out why those exceptions would be thrown (though in these cases it's pretty self-evident in most cases).

But, like I said, I'm willing to see if I'm ever really bothered about this again given that so rarely do I use quick info for exception info anyway. I usually use it to remind me of the names of the arguments being passed). If there were a mapping from particular exceptions to the parameters that cause them then this information _might_ be useful in signature help. I'm not entirely sure knowing what exceptions a method throws is useful at all, let alone without reasons. It's not like our languages have checked exceptions and it's also not like I often handle exceptions--usually I just let them crash the process. File I/O is probably the one place where exceptions happen in very non-exceptional cases.

So in this particular case quick info is taking up a lot of space but I'm not entirely clear why. It should either add more value or not take up so much space.

**Comment by AraHaan** [comment]
Alright so I got an idea, what if when it is detected that there is over like 5 exceptions in a method that it shows the arrow thing like how it does when it finds method overloads? Just this arrow thing right on the other side of the method overload one and is labeled "Exceptions x/?????" where x is the current page viewed on the exception part and ????? being the total pages. And then for 5 exceptions or less do not show the arrow unless the bounds of the quick info seems to big?

**Comment by CyrusNajmabadi** [comment]
@AraHaan i think we would take a PR that provided that functionality :)

**Comment by AraHaan** [comment]
@CyrusNajmabadi so to install the local copy of roslyn onto my instance of vs when I need to test it what vsix would I have to run?

**Comment by CyrusNajmabadi** [comment]
hey @AraHaan if you open Roslyn.sln you'll see: 

![image](https://user-images.githubusercontent.com/4564579/79374361-21b3fb80-7f0c-11ea-8ec1-de235d8934e8.png)

You can ctrl-f5 to build this roslyn and then launch a new VS with those roslyn bits installed.  The bits will be installed in an isolated fashion, so it won't impact your main development experience.  i.e. if you introduce a crash, you can still develop :)

**Comment by AraHaan** [comment]
Alright made progress. However I need some help filtering some more possible xml so that text gets colored right in the ide.

![BB930734-6977-4AA9-A5E7-9E7D19E6AF25](https://user-images.githubusercontent.com/15173749/79466819-1542a800-7fcb-11ea-825a-29276f79ed11.jpeg)
![0CBB08E9-7141-4EB1-8DF9-7C948755799D](https://user-images.githubusercontent.com/15173749/79466842-1c69b600-7fcb-11ea-8749-c2192951b163.jpeg)

**Comment by AraHaan** [comment]
I am starting back progress on this to try to bring this into VS2022 *hopefully*. Until then I will need to decide on a few questions as well:

- Should I limit displaying text for each exception to a single line (strip all ``<br>`` and ``<para>`` inside of them)?
- Should I limit the text to 80 characters or 100 characters per line?
- Should I limit the number of exceptions shown but somehow have them be paged out like you can see arrows for when the QuickInfo displays for methods that has overloads that exists? If so how many? 5? 7? 9? 10? More than 10?

I think that above just about covers it.

**Comment by AraHaan** [comment]
So, after a year of having the code, I realized 1 major part of it (paginating the exception reasons) would have to be done by the IDE team itself as the code to it is not accessible on the roslyn side.

With that said, I think all I have to do now is fixup the tests somehow to make them work as intended.

**Comment by janseris** [comment]
This featue would be very welcome.
Considering the posts here about that for some developers, the info would be too large etc - the amount info is shown (and if shown at all) could be an option in the Settings so that people who don't want it can disable it (or disabled by default)?

E.g:
Method documentation quick preview "Exception description" feature - Enable/Disable
- Exception description length: low, medium, full
?

**Comment by AraHaan** [comment]
I would prefer it to have exception description not be disableable, but an setting on how many can be placed on each page of them when they are paginated when the # of exceptions is > the amount they configured (between 5 and 9 being the only valid range).

**Comment by janseris** [comment]
@jinujoseph  any updates on this?

**Comment by AraHaan** [comment]
I been working locally on unit tests.

**Comment by fissssssh** [comment]
Hope to see this feature soon, it's really useful!

**Comment by AraHaan** [comment]
It still has work that needs done and some further planning to do to determine if truncation is needed or not.

Overall currently I think these are needed to be implemented in the settings of visual studio:

- truncation of each exception documentation, default limit is null to represent no limit.
- max number of exceptions per page of exception documentations, default 6, Min: 1, Max: 9

**Comment by ChrisOwensKPMG** [comment]
Would be really good to see this, I find it hard to believe this hasn't been a feature for the past 8 years as it seems obvious to me this is how it should work!

**Comment by CyrusNajmabadi** [comment]
@ChrisOwensKPMG it's in the backlog/help-wanted, and no one has found it important enough to contribute.  If you're interested, that would be great :)

**Comment by janseris** [comment]
@CyrusNajmabadi 
In fact it has been already implemented but not reviewed from the Microsoft Team.
https://github.com/dotnet/roslyn/pull/56162
You can see this on the issue info on the right in the start of this page.

**Comment by janseris** [comment]
> That PR hasn't seen movement in a year. If you'd like to work on it though, the other contributor may be interested in working on it with you :)

This PR is pending review. There is nothing to work on from the community.

**Comment by CyrusNajmabadi** [comment]
There has been no communication from the author indicating any work is happening on it, or if they're looking for review or anything else. If they come back (or someone has time to pick this up), it could potentially go in.

**Comment by AraHaan** [comment]
Yeah, I been having some challenges with it somewhat as well along with some other challenges in real life as well.
